Updated find_broken_builds to save build_broken flag and reason

orangejenny opened the pull request:
## Technical Summary
https://dimagi-dev.atlassian.net/browse/SAAS-14326

Since all builds were broken during the time the bug was live, this adds a new check function that flags all builds.

`build_broken` and `build_broken_reason` aren't surfaced in the UI, so this is only for historical purposes.

## Safety Assurance

### Safety story
Tested locally. This is a low risk change to a low risk command.

### Automated test coverage

None

### QA Plan

None


### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Updated find_broken_builds to save build_broken flag and reason

Jtang-1 commented on corehq/apps/app_manager/management/commands/find_broken_builds.py:
```diff
@@ -133,7 +146,7 @@ def get_build_ids(start, end):
+def find_broken_builds(checker, builds_ids, reason=None):
     for build in iter_docs(Application.get_db(), builds_ids):
         error = checker(build)
         if error:
```
Will this always evaluate to true if `all_builds` is passed as the check function since `all_builds` always returns the `error`?

Jtang-1 replied:
Oh, I may have misunderstood the purpose of this change. The checker is intended to be used to find all builds that are built within a time frame or be a certain build_id such that it can be flagged, so this is what we want.

What confused me is that in the general case, the returned results won't all be broken builds, which conflicts with the function name `find_broken_builds`

Jtang-1 commented on corehq/apps/app_manager/management/commands/find_broken_builds.py:
```diff
@@ -144,3 +157,8 @@ def find_broken_builds(checker, builds_ids):
+            if reason:
+                build = Application.wrap(build)
+                build.build_broken = True
+                build.build_broken_reason = reason
```
nitpick: Should the [comment](https://github.com/dimagi/commcare-hq/blob/6c3d20ca9dc4d3cee9d7bcffc4b7afd487d8333a/corehq/apps/app_manager/models.py#L4245-L4248) related to build_broken_reason be updated to reflect this usage?

orangejenny replied:
yes, updated

joxl commented on corehq/apps/app_manager/management/commands/find_broken_builds.py:
```diff
@@ -44,6 +44,10 @@ def form_filter_error(build):
             prev_form_filter = form_filter
 
 
+def all_builds(build):
```
Perhaps this function could be named something more descriptive regarding what it's for? Alternatively a docstring would be helpful.

```suggestion
def fail_all_builds(build):
```

joxl reviewed (commented):
> What confused me is that in the general case, the returned results won't all be broken builds, which conflicts with the function name `find_broken_builds`

I also share this sentiment.  I wasn't able to effectively review this change until I reviewed the full original source in order grok what is even going on here. The diff in this PR alone is quite confusing to someone not intimately familiar with this tool due to the contradictory names @Jtang-1 mentioned above.

This change seems like a bit of an abuse of a manage command named `find_broken_builds`. With these changes, the command is much less of a "checker", and much more of a "flagger". When running it with the new `all` argument, there is no "checking" going on, rather the caller of the command seems to have predetermined that _all_ builds are broken and is using the command to flag them as such.  Perhaps this command could be renamed?

orangejenny committed:
Renamed command

orangejenny committed:
Switched from reason as argument to reason as user input

Long string arguments taht need to be quoted are annoying, plus now that
the command is named 'flag', flagging with a reason ought to be part of
the standard workflow.

orangejenny committed:
Updated check funciton to be optional

This is clearer than having a 'checker' that just returns all builds

orangejenny committed:
Updated help

orangejenny committed:
Updated comment

orangejenny commented:
Added a few clarifying commits, renaming the command and making its default behavior to flag builds instead of only identifying them.

Found [this old wiki](https://confluence.dimagi.com/display/saas/Flagging+broken+builds) relating to this comment. Apparently there used to be a UI dealing with flagged builds. I'll update that page once this PR is settled.

joxl commented:
> flagging with a reason ought to be part of the standard workflow.

:star: 

Nice find with the wiki!

orangejenny commented:
Ran in a private release on prod and india, details in ticket.

Merging without tests - there's no test coverage and this has already been run.
